pkg/nsx/services/firewall.go

```diff
@@ -281,18 +363,209 @@ func (service *SecurityPolicyService) addOperatorIfNeeded(expressions *data.List
+func (service *SecurityPolicyService) updateExpressionsMatchLables(matchLabels map[string]string, memberType string, expressions *data.ListValue) {
 	for k, v := range *util.NormalizeLabels(&matchLabels) {
 		service.addOperatorIfNeeded(expressions, "AND")
 		expression := service.buildExpression(
-			"Condition", "SegmentPort",
+			"Condition", memberType,
 			fmt.Sprintf("%s|%s", k, v),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(expression)
 	}
 }
 
+// NSX understand the multiple values w.r.t a key in a joined string manner
+// this function iterates over input matchExpressions LabelSelectorRequirement
+// with same operator and Key, and merges them into one and values to a joined string
+// e.g. {'NotIn': [{'k1': ['a1', 'a2', 'a3']}, {'k1': ['a2', 'a3', 'a4']}]}
+//   => {'NotIn': [{'k1': 'a1, a2, a3, a4'}]
+func (service *SecurityPolicyService) mergeSelectorMatchExpression(matchExpressions []metav1.LabelSelectorRequirement) *[]metav1.LabelSelectorRequirement {
+	mergedMatchExpressions := make([]metav1.LabelSelectorRequirement, 0)
+	isMerged := make([]bool, len(matchExpressions))
+	var meregedSelector metav1.LabelSelectorRequirement
+	var i, j = 0, 0
+	var isFound = false
+
+	for i = 0; i < len(matchExpressions); i++ {
+		if isMerged[i] {
+			continue
+		}
+		for j = i + 1; j < len(matchExpressions); j++ {
+			if matchExpressions[i].Operator == matchExpressions[j].Operator &&
+				matchExpressions[i].Key == matchExpressions[j].Key {
+				isMerged[j] = true
+				if !isFound {
+					meregedSelector.Values = util.MergeStr(matchExpressions[i].Values, matchExpressions[j].Values)
+					meregedSelector.Operator = matchExpressions[i].Operator
+					meregedSelector.Key = matchExpressions[i].Key
+					isFound = true
+				} else {
+					meregedSelector.Values = util.MergeStr(meregedSelector.Values, matchExpressions[j].Values)
+				}
+			}
+		}
+
+		if !isFound {
+			mergedMatchExpressions = append(mergedMatchExpressions, matchExpressions[i])
+		} else {
+			mergedMatchExpressions = append(mergedMatchExpressions, meregedSelector)
+			isFound = false
+		}
+	}
+	return &mergedMatchExpressions
+}
+
+// Todo: check limit for mixed namespaceSelector, podSelector, VMSelector
+// Given NSX currently doesn't support 'In' LabelSelector, to keep design simple,
+// only allow just one 'In' LabelSelector in matchExpressions with at most of five values in it.
+func (service *SecurityPolicyService) validateSelectorExpressions(matchLabels map[string]string,
+	matchExpressions []metav1.LabelSelectorRequirement, member_type string) (int, error) {
+	var exprInCount = len(matchLabels)
+	var mexprInOpcount = 0
+	var mexprInValueCount = 0
+	var err error = nil
+	var errorMsg string = ""
+
+	for _, expr := range matchExpressions {
+		if expr.Operator == metav1.LabelSelectorOpIn {
+			mexprInOpcount++
+			mexprInValueCount += len(expr.Values)
+			if mexprInOpcount > MaxMatchExpressionInOp {
```

Comment:
Why we only support 1 `In` operator?
I think we could also support below case, right?
    matchExpressions:
      - {key: k1, operator: In, values: [v1]}
      - {key: k2, operator: In, values: [v2]}

Reply:
This limit design for In operator is following NCP's design. but Since nsx limit, Operator In could get more complex. 
There is a lot of case to handle since the code need to make a cross product, and then make a OR operator with each member of product.  so for the example above, it would be (k1 == v1 AND k2 == v2) OR  {....}.
it easier to lead to create a big nested expressions. Ideally, we can use cross product number to check, if this, we can support more cases for operator IN, but not fully